src/util.c

```diff
@@ -572,22 +598,11 @@ int d2string(char *buf, size_t len, double value) {
         else
             len = snprintf(buf,len,"0");
     } else {
-#if (DBL_MANT_DIG >= 52) && (LLONG_MAX == 0x7fffffffffffffffLL)
-        /* Check if the float is in a safe range to be casted into a
-         * long long. We are assuming that long long is 64 bit here.
-         * Also we are assuming that there are no implementations around where
-         * double has precision < 52 bit.
-         *
-         * Under this assumptions we test if a double is inside an interval
-         * where casting to long long is safe. Then using two castings we
-         * make sure the decimal part is zero. If all this is true we use
-         * integer printing function that is much faster. */
-        double min = -4503599627370495; /* (2^52)-1 */
-        double max = 4503599627370496; /* -(2^52) */
-        if (value > min && value < max && value == ((double)((long long)value)))
-            len = ll2string(buf,len,(long long)value);
+        long long lvalue;
+        /* Integer printing function is much faster, check if we can safely use it. */
+        if (double2ll(value, &lvalue))
+            len = ll2string(buf,len,lvalue);
         else
-#endif
             len = snprintf(buf,len,"%.17g",value);
```

Comment:
@oranagra I wrote a unittest for `zzlInsert()`, and found that it does not save memory.
You mentioned in the top comment that using `snprintf` might result in storing strings, but I added logs and found that if the value is long long, the `buf` does not output a string like `*.00000`.
So before this pr, zset was already storing integer.

```c
len = snprintf(buf,len,"%.17g",value);
printf("value, %lf, buf: %s\n", value, buf);

# output
value, 14503599627340496.000000, buf: 14503599627340496
```

And I used the test below, and found that the two branches(unstable and this pr) ran in essentially the same amount of time consume.
Please point out if I am wrong.

```c
long long start = ustime();
size_t memory_used = zmalloc_used_memory();
unsigned char *lp = lpNew(0);
for (long long i = 14503599627370496; i > 14503599627370496 - 30000; i--) {
    sds ele = sdscatfmt(sdsempty(), "key%ld", i);
    lp = zzlInsert(lp, ele, i);
    sdsfree(ele);
}
printf("consume: %lld, memory_used: %zu\n", ustime() - start, zmalloc_used_memory() - memory_used);
printf("listpack length: %ld, bytes: %zu\n", lpLength(lp), lpBytes(lp));
```

Reply:
@sundb i didn't refer to the `.0000` suffix, i referred to the fact is stores a string (characters) in the listpack, instead of an integer.
i.e. the score of 1000 will take 5 bytes (4 characters plus one byte for listpack encoding header), but only 2 bytes as integer encoding.
am i missing something or did you?
i also think it'll be faster, since we don't bother to convert the integer to string and later back from string.

Reply:
I feel like you missed https://github.com/redis/redis/blob/b8eb2a73408fa3b8845760857dd6fcccb62107fe/src/listpack.c#L331
When score is int64, it will be automatically stored as integer, not string.

Reply:
Indeed I missed it. It means that my PR doesn't affect memory usage, or changes the actual contents of the listpack. But it should still save some cpu time, as it avoids the conversion to string and back to int. 
Or am I missing anything else?

Reply:
Looking at the code, it must be able to bring improvement, my test cpu consumption time fluctuates greatly, I may need to modify the test to verify it more accurately.

Reply:
Some to think of it, since my PR makes sure to add scores that are integer encoded as integers, maybe it's a good idea to tip the listpack code not to [try](https://github.com/redis/redis/blob/b8eb2a73408fa3b8845760857dd6fcccb62107fe/src/listpack.c#L331) the conversion on its side (another cpu saving)

Reply:
Remove the integer detection from lpInsertString and let the caller detect it and uae the rifht function?

Reply:
@oranagra @zuiderkwast If we remove this, it will make code more complicated, all the places where using `lpInsertString`, `lpPrepend`, `lpReplace` need to determine in advance whether the string is an integer or not.
BTW, `lpInsertInteger`, `lpPrependInteger`, `lpAppendInteger` do not reach `lpStringToInt64()`,

Reply:
Yes, if we already know it's an integer (lpInsertInteger), we obviously don't try to convert it to an integer. But in case the score was double, and we converted it to string, this code will try once again to check if it can be converted to int. 

I don't think we should remove the detection from lpInsertString, but maybe find some way to hint listpack to skip it in some cases. 

Another alternative is to add an lpInsertDouble, which will take care of all of this, and avoid using d2string in zset

Reply:
`lpInsertDouble` +1